asab/api/discovery.py

```diff
@@ -33,81 +54,170 @@ async def locate(self, instance_id: str = None, service_id: str = None) -> list:
+	async def _locate(self, instance_id: str = None, **kwargs) -> typing.Set[typing.Tuple]:
 		"""
-		Locates service instances based on their instance ID or service ID.
+		Locates service instances based on their identifiers.
 
 		:param instance_id: The unique identifier for a specific instance of a service
 		:type instance_id: str
-		:param service_id: The ID of the service to locate
-		:type service_id: str
 		:return: a list of tuples containing the server name and port number of the located service(s).
 		"""
-		if instance_id is None and service_id is None:
-			L.warning("Please provide instance_id, service_id, or appclass to locate the service(s).")
-			return
+		res = set()
 
-		instances = await self.get_advertised_instances()
-		if len(instances) == 0:
-			L.warning("No instances available.")
-			return
-		res = []
-		for instance in instances:
-			if service_id is not None:
-				if service_id != instance.get("service_id", "").lower():
-					continue
+		if instance_id is not None:
+			locate_params = {"instance_id": instance_id}
+		elif len(kwargs) > 0:
+			locate_params = kwargs
+		else:
+			L.warning("Please provide instance_id, service_id, or other custom id to locate the service(s).")
+			return res
 
-			if instance_id is not None:
-				if instance_id != instance.get("instance_id", "").lower():
-					continue
+		if self.AdvertisedCache is None:
+			L.warning("Discovery service is not yet ready to locate services.")
+			return res
 
-			web = instance.get("web")
-			host = instance.get("node_id", instance.get("host"))
+		if len(self.AdvertisedCache) == 0:
+			L.warning("No instances to discover.")
+			return res
+
+		for id_type, ids in self.AdvertisedCache.items():
+			if id_type in locate_params:
+				if locate_params[id_type] in ids:
+					res = res | ids[locate_params[id_type]]
 
-			if web is None or host is None:
-				continue
-			for i in web:
-				try:
-					ip = i[0]
-					port = i[1]
-				except KeyError:
-					L.error("Unexpected format of 'web' section in advertised data: '{}'".format(web))
-					return
-				if ip not in ("0.0.0.0", "::"):
-					continue
-				res.append((host, port))
 		return res
 
+	def discover(self) -> typing.Dict[str, typing.Dict[str, typing.Set[typing.Tuple[str, int]]]]:
+		return self.AdvertisedCache
 
 	async def get_advertised_instances(self) -> typing.List[typing.Dict]:
 		"""
+		This method is here for backward compatibility. Use `discover()` method instead.
 		Returns a list of dictionaries. Each dictionary represents an advertised instance
 		obtained by iterating over the items in the `/run` path in ZooKeeper.
 		"""
+		LogObsolete.warning("This method is obsolete. Use `discover()` method instead.")
 		advertised = []
 		async for item, item_data in self._iter_zk_items("/run"):
-			data_dict = json.loads(item_data)
-			data_dict['zookeeper_id'] = item
-			advertised.append(data_dict)
+			item_data['zookeeper_id'] = item
```

Comment:
`zookeeper_id` ... find a better name, don't disclose the technology name.

I.e. `ephemeral_id` ...